tool/assetfinder.go

```diff
@@ -0,0 +1,96 @@
+package tool
+
+import (
+	"fmt"
+	"sync"
+	"time"
+    "os"
+
+	"github.com/fatih/color"
+	assetfinder "github.com/spiral-sec/assetfinder/scanner"
+
+	"github.com/CMEPW/Yelaa/helper"
+)
+
+
+type fetchFn func(string) ([]string, error)
+
+type AssetfinderConfig struct {
+    outfile string
+    functions []fetchFn
+}
+
+type Assetfinder struct {
+    opts *AssetfinderConfig
+    outfile string
+}
+
+func (a *Assetfinder) Info(url string) {
+    color.Cyan("Running Assetfinder on %s", url)
+}
+
+func (a *Assetfinder) Configure(c interface{}) {}
+
+
+func (a *Assetfinder) Run(url string) {
+    var wg sync.WaitGroup
+    outfile := helper.YelaaPath + "/assetfinder.txt"
+    functions := []fetchFn{
+        assetfinder.CertSpotter,
+        assetfinder.HackerTarget,
+        assetfinder.ThreatCrowd,
+        assetfinder.CrtSh,
+        assetfinder.Facebook,
+        assetfinder.VirusTotal,
+        assetfinder.FindSubDomains,
+        assetfinder.Urlscan,
+        assetfinder.BufferOverrun,
+    }
+
+    rl := assetfinder.NewRateLimiter(time.Second)
+    out := make(chan string)
+
+    for _, f := range functions {
+        wg.Add(1)
+        fn := f
+
+        go func () {
+            defer wg.Done()
+
+            rl.Block(fmt.Sprintf("%#v", fn))
+            names, err := fn(url)
+
+            if err != nil {
+                return
+            }
+
+            for _, n := range names {
+                n = assetfinder.CleanDomain(n)
+                out <- n
+            }
+        }()
+    }
+
+    go func() {
+        wg.Wait()
```

Comment:
Wait in a goroutine?

Reply:
That's how it was implemented in the original Assetfinder codebase, I decided to leave it in case it slowed heavily the program. I can benchmark though if you want

Reply:
We wait to be sure that all workers are done before closing the channel

Reply:
The wait won't block the function so I don't think it's a good thing to fork the process using a goroutine. I don't see the benefits.

Reply:
Ok, I'll remove the goroutine call

Reply:
![without-fork](https://user-images.githubusercontent.com/29751352/169716190-74e799da-07a9-46e9-ac70-c1a68b04c681.png)

It seems that assetfinder deadlocks if we don't do the wait in a fork. I tried again with it, to check if my connection was the issue and it spit out the full scan for the two domains in less than 4 minutes.

Reply:
Not really fan but I don't have time to investigate on it. So I'm okay with the changes.

Reply:
I will check it out in the weekend because that makes me curious, I will merge on sunday if i don't find a sufficient answer. if i do, i'll come back here and tell you